### ISSUE TITLE
Module Overrides for Guice bundle

### ISSUE THREAD

**Issue by pandaadb** [xvik/dropwizard-guicey#23]
Hi,

This is not really an issue but more of a feature request/suggestion. I wanted to see how you feel about that. 

I have the following scenario: 

I have my application class ad i discover my modules for instantiation. Same with all extensions, so essentially i just wire things together and they work :) 

However, with testing, sometimes you have to supply mocks, or supply different beans etc. This is for integration testing where you'd want the entire server to be able to start up. 

I read about Modules.overwrite in guice. I think this could be an interesting addition (i couldn't find any info on that so I am not sure if you have something like that already). For example I would imagine something like this:

* create app rule 

RULE = new DropwizardTestSupport<TestConfiguration>(Tester.class, "test");

we'd have to overwrite the test support I suppose for this to work. 

* Then have maybe a method: 

RULE.override(FromModule, ToModule) 

* And finally start up the server 

RULE.before() 

The issue this would solve for me is that  I wouldn't have to override my main application class for test, where I can then provide manually which modules/bundles to be installed. I could simply start up the default config and override the correct impl. 

What might be even cooler would be to maybe be able to override specific beans. Maybe in form of a test-annotation on the bean? So, say I have a provider that provides a client: 

@TestOverride(key="customClient")
public class MyProvider implements Provider<Client> { ...} 

Then in the RULE, we could have: 

RULE.override(key, bean) 

This way, one would not even have to override modules and/or application classes.

The main danger with overrides that I see is that things may go untested. For example, say one has a module that provides a certain impl. Since the tests override the module and the app class to provide the correct mocked objects in the test, one might overlook that configuration. So Dev 1 overrode Module 1. Dev 2 changes a bean in Module 1 thinking that the existing integration tests would test this change properly. However, since Dev 1 had to overwrite the entire Module 1, the later change is never tested because it is not represented in the Module change. 

By being able to overwrite a specific bean only, the configuration for the test would be the same as the one for production with the additon of mocked objects. 

What do you think? 

-- Artur

**Comment by xvik** [comment]
Hi,

Yes, that is exactly what I'm thinking of (and not so long ago @Ignas was asking about this too in gitter). It is definitely the next thing I will do. 

But I'm not sure how will it look like. I was close to something like this in previous releases (and, partially, detailed diagnostic was done for such complex disable/replace cases in the future): internal api allows disabling anything, but public api still allows to disable just an installer. It was urgent to release dw 1.0 version so I finally postpone this to think better later.

So Internally guicey is 100% ready for any kind of overriding staff: it knows all extensions before starting guice and can do anything with them. 
Generally speaking, overriding could be handful not just for tests, but for an application itself (to be able to very flexibly tune other bundles or be able to fix bugs in it).

I'm not sure how it will be done. Most likely there will be new configuration methods like disableExtension, disableBundle etc or replaceExtension, replaceBundle (maybe some mix of them.. not sure for now).

Also, It is possible to introduce new extension point like ExtensionsPreProcessor which will be able to disable/replace/add whatever (something like registration listener). With it, for example, it would be possible to disable all extensions in some package (or implementing interface or whatever..). Maybe this new ability will be available only for tests (most likely because it's way to powerful for common usage).

Overall, I can assure you that extended tests support is nb 1 feature for the next release. 
But, right now I have no time for it (you can see this by my github commit activity). 
Most likely, I will start this only at the end of December (and release will definitely be only next year).

Moreover, I have started two more projects (not public for now) and maybe I will release at least one of them before new guicey version. It should be common guicey extensions (various integrations and helpers). Not 100% sure about this, but most likely (because I want to move out this staff from current projects to something re-usable :) )

**Comment by pandaadb** [comment]
That sounds nice 

I am currently trying to work around this using Guice. However it does not seem like guice has a hook to change creation of instances (e.g. to switch them out). I got a little progress using reflection and trying to overwrite the result of an instantiation (provision) as such: 

```
bindListener(Matchers.any(), new ProvisionListener() {
							
							@Override
							public <T> void onProvision(ProvisionInvocation<T> provision) {
								Binding<T> binding = provision.getBinding();
								
								TypeLiteral<T> typeLiteral = binding.getKey().getTypeLiteral();
								
								OverrideConfig overrideConf = overrides.get(typeLiteral.getRawType());

								
								
								if(overrideConf != null) {
									
									if(overrideConf.name != null) {
										Key<T> key = binding.getKey();
										Annotation annotation = key.getAnnotation();
										if(annotation == null) {
											return;
										}
										String name = null;
										
										if(annotation instanceof Named) {
											Named n = (Named) annotation;
											name = n.value();
										}
										
										if(annotation instanceof com.google.inject.name.Named) {
											com.google.inject.name.Named n = (com.google.inject.name.Named) annotation;
											name = n.value();
										}
										
										if(name == null || !(name.equals(overrideConf.name))) {
											return;
										}
										
									}
									
									// execute the provisioning to get guice happy
									T res = provision.provision();
									
									// hack guice to enable overrides for simple types
									Field field;
									try {
										field = provision.getClass().getDeclaredField("result");
										field.setAccessible(true);
										field.set(provision, overrideConf.replacement);
									} catch (NoSuchFieldException | SecurityException | IllegalArgumentException | IllegalAccessException e) {
										e.printStackTrace();
									}
								}
							}
						});
```

This however only worked for actual objects, it ignores instances (so i imagine it would only work on bindings of the form: bind(x.class).to(xImpl.class)) 

I now wonder if the better way for now would be to use a custom InjectorFactory. That way, i would have the power to overwrite only that method in my main class, and then simply use Modules.overwrite and specify my overwriting beans in the new module. I tried ripping off the overwrite method, however it is very complex and i do not quite understand  what is happening in there :D

**Comment by xvik** [comment]
Yes, ProvisionListener is not suitable because it ignores untargeted bindings. TypeListener could be used  (it has limitations with providers but generally works) but it would be a nightmare replacing all instances manually with reflection.

I like your idea about overriding InjectorFactory. Indeed this way you will be able to use Modules.override() and so could override bindings. As I tell you in the previous issue, all extensions are bound inside InstallerModule (or by an installer, but to binder from this module) and so should be available for overriding.
I don't know what limitations this approach could have but looks very promising. 

Honestly, I don't understand what for are you trying to "ripping off the overwrite method". It should work as is without hacks (in theory). 

Please tell me if you succeed or not with this approach.

**Comment by pandaadb** [comment]
Hi,

yes I think I am going down the route of using Modules.override. It looks the cleanest and simplest solution to me.

By  "ripping off the overwrite method" I meant that I looked into how guice is implementing it, in terms of being able to extract the right portion of that and simply use it :)

The InstallerModule seems not 100% suitable for that? I might be misunderstanding the class, but I think it would only offer the possibility of overriding extension classes, not Module bound classes. So i would have to declare all my beans as extensions rather than in modules? 

Thanks! 

Artur

**Comment by xvik** [comment]
Hi,

No-no I didn't mean that. I point to installer module just to show that extension bindings will be created by this module and so could be overridden. 
If I understand guice override feature correctly, all you have to do inside custom injection factory is:

```java
Guice.createInjector(state, 
               Modules.override(modules)
                            .with(new MyOverridingModule())));
```

Where MyOverridingModule will contain required overrides. All other bindings will be used from production modules.

Of course, such factory must be smart enough to know when to override (probably with some system property or env variable). Or application itself could check some flag and configure appropriate injector factory.

**Comment by pandaadb** [comment]
Yes, this is the way I wanted to go down to. Basically expose the creation of the injectorfactory so that the test can override this and pass the correct override module in.

I agree there should be some layer of safety there so that noone gets the idea of overriding things when it is not in test. But to be honest, if someone wanted to override dependencies, they would find a way anyway :) I am hoping for "common sense" again - let's see how long that last before I regret that faith :)

**Comment by tmszdmsk** [comment]
Wow, it's great to hear that you're thinking about that. I resolved "overriding modules in test" issue in the company I work in but it isn't too generic. I simply split the Application into ApplicationDefinition and ApplicationCreator.  ApplicationDefinition contains information about Dropwizard bundles and Guicey modules/extensions/installers you want to have in your application. ApplicationCreator part just reads that 'definition' and creates Application instance based on that. I also added junit Rule that works similarly to DropwizardAppRule but before it creates the application instance it allows to provide Module passed to Modules.override. It isn't too flexible but it's enough for our needs. 

Problems we encountered:
-  in that way you can't override bindings from FeatureInstallers, there is a plan to not only allow to override modules, but also disable features.
- the same for .enableAutoConfig()
- it breaks (Configuration|Environment|Bootstrap)AwareModules since that overriding module doesn't implement those interfaces.

Probably all of them can be resolved when making changes in guicey internals, but maybe it will highlight potential issues.

**Comment by xvik** [comment]
@tmszdmsk thank you for feedback! Right now, the best solution for mocking is @pandaadb's suggestion to use custom InjectorFactory. I tried this approach recently and can assure that it works. The main idea is that InjectorFactory receives the list of modules and able to apply Modules.override on them.

The simplest implementation:
```java
public class CustomInjectorFactory implements InjectorFactory {
     private static ThreadLocal<Modules[]> modules = new ThreadLocal<>();

     @Override
     public Injector createInjector(Stage stage, Iterable<Module> modules) {
          Modules[] customModules = modules.get();
          modules.remove();
          return Guice.createInjector(stage, customModules == null ? modules 
                               : Lists.newArrayList(Modules.override(modules).with(customModules)) ) 
     }

     public static void override(Module... modules) {
          customModules.set(modules);
     }
}

// custom factory registration 
GuiceBundle.builder().injectorFactory(new CustomInjectorFactory())
```

In tests we could simply do: 
```java
CustomInjectorFactory.override(new MyOverridingsModule());
```

Of course, in some cases, it's required to drop entire bundles/features/whatever and, in such cases, only something like @tmszdmsk suggestion will help.

Eventually, all this will be supported.. just need to think about all cases and carefully implement them. 

Regarding problems:
>in that way you can't override bindings from FeatureInstallers, there is a plan to not only allow to override modules, but also disable features. the same for .enableAutoConfig()

The method above allows overriding all bindings (because it overrides all possible application bindings). Disabling bundles/features will be supported. But even now it's possible to disable installer (using custom bundle + properties bundles lookup) which will disable all features installed by this installer (but only if features resolved with classpath scan). 

> it breaks (Configuration|Environment|Bootstrap)AwareModules since that overriding module doesn't implement those interfaces.

Very good point.

**Comment by xvik** [comment]
New concept `GuiceyConfigurator` which receives GuiceBundle's builder (just after its creation: `GuiceBundle.builder()`). 

Added support for disabling bundles, extensions and modules in builder:
* `builder.disableBundles()`
* `builder.disableExtensions()`
* `builder.disableModules()`

Added generic predicate-based diable method in builder: `builder.disable(predicate)`. Example usage:

```java
import static ru.vyarus.dropwizard.guice.module.context.Disables.*

 builder.disable(installer()
             .and(registeredBy(Application.class))
             .and(type(SomeInstallerType.class).negate());
```
or

```java
import static ru.vyarus.dropwizard.guice.module.context.Disables.*

 builder.disable(type(MyExtension.class,
                      MyInstaller.class,
                      MyBundle.class,
                      MyModule.class));
```

Added direct support for guice bindings overriding: `builder.overrideModules()`. All registered overriding modules will be applied with guice's `Modules.override` (override existing bindings from simple `modules()` and simply add all new bindings). Overriding modules are also processed with *Aware interfaces to inject configuration, bootstrap and environment objects.

**For junit** there is new rule to declare configurator: `GuiceyConfiguratorRule`, but it must be used with `RuleChain` to preserve order:

```java
static GuiceyAppRule RULE = new GuiceyAppRule<>(App.class, null);

@ClassRule
public static RuleChain chain = RuleChain
        .outerRule(new GuiceyConfiguratorRule((builder) -> builder.modules(...)))
        .around(RULE);
```
[GuiceyAppRule example](https://github.com/xvik/dropwizard-guicey/blob/master/src/test/groovy/ru/vyarus/dropwizard/guice/test/configurator/unit/ConfigurerGuiceyAppTest.java), [DropwizardAppRule example](https://github.com/xvik/dropwizard-guicey/blob/master/src/test/groovy/ru/vyarus/dropwizard/guice/test/configurator/unit/ConfigurerDwAppTest.java), [Base configurator definition example](https://github.com/xvik/dropwizard-guicey/blob/master/src/test/groovy/ru/vyarus/dropwizard/guice/test/configurator/unit/MultipleConfiguratorsTest.java) (not so universal, but the only way.. better then nothing)

**For spock**, `configurators` property was added to `@UseGuiceyApp` and `@UseDropwizardApp`:

```java
@UseGuiceyApp(value = App, configurators = Conf)

static class Conf implements GuiceyConfigurator {
        @Override
        void configure(GuiceBundle.Builder builder) {
            builder.modules(new XMod())
        }
    }
```
And new extension `@UseGuiceyConfigurator` could be used to declare base configurator in the base class:

```java
@UseGuiceyConfigurator(Conf)
```
[UseGuiceyApp example](https://github.com/xvik/dropwizard-guicey/blob/master/src/test/groovy/ru/vyarus/dropwizard/guice/test/configurator/GuiceyAppConfiguratorAttributeTest.groovy), [UseDropwizardApp example](https://github.com/xvik/dropwizard-guicey/blob/master/src/test/groovy/ru/vyarus/dropwizard/guice/test/configurator/DwAppConfiguratorAttributeTest.groovy), [base configurator definition example](https://github.com/xvik/dropwizard-guicey/blob/master/src/test/groovy/ru/vyarus/dropwizard/guice/AbstractTest.groovy)

**Please tell me if you see any potential problems or how to improve this**
